File: src/route_decider.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

const double SHORT_ROUTE = 598.34;
const double LONG_ROUTE = 698.26;
const int MAX_SPEED = 20;

struct LatestMessage {
    string edge;
    double speed;
};

vector<string> edges_short = {"main1"};
vector<string> edges_long = {"side1", "side2", "side3"};

bool continue_on_route(vector<LatestMessage> latest, double short_speed, double long_speed)
//Add an if-check for whether they are before at the start so they actually can decide the route
{
    for (auto message : latest)
    {
        if (find(edges_short.begin(), edges_short.end(), message.edge) != edges_short.end() && message.speed < short_speed)
        {
            short_speed = message.speed;
        }
        if (find(edges_long.begin(), edges_long.end(), message.edge) != edges_long.end() && message.speed < long_speed)
        {
            long_speed = message.speed;
        }
    }

    return short_speed * SHORT_ROUTE < long_speed * LONG_ROUTE;
}

int main()
{
    // Example usage of continue_on_route function
    vector<LatestMessage> latest_messages = {{"main1", 15}, {"side1", 10}, {"side2", 20}};
    bool continue_on_short_route = continue_on_route(latest_messages, MAX_SPEED, MAX_SPEED);
    cout << "Continue on short route? " << (continue_on_short_route ? "Yes" : "No") << endl;

    return 0;
}

File: src/route_decider.py
"""A code example of how the deciding of routes could be handled, probably want this in C++"""

#Defines for route lenght and max speed on them
SHORT_ROUTE = 598.34
LONG_ROUTE = 698.26
MAX_SPEED = 20

#Might want to remove some of them since cars at the end of each route is not interesting
edges_short = ["main1"]
edges_long = ["side1", "side2", "side3"]

#Default route is short route, so this function returns a boolean of whether the car should
# continue of the route or switch to the side road due to congestion. Latest is a list of all the
# latest messages, one from each car where the car is the key. 
def continue_on_route (latest, short_speed, long_speed):
    """Return boolean if it should continue on the route"""
    for key in latest:
        if key.get("edge") in edges_short and key.get("speed") < short_speed:
            short_speed = key.get("speed")
        if key.get("edge") in edges_long and key.get("speed") < long_speed:
            long_speed = key.get("speed")

    return short_speed * SHORT_ROUTE < long_speed * LONG_ROUTE
